internal_worker_base.go

```diff
@@ -121,18 +133,128 @@ func (bw *baseWorker) Start() {
 		return
 	}
 
-	// Launch the routines to do work
-	for i := 0; i < bw.options.routineCount; i++ {
+	for i := 0; i < bw.options.pollerCount; i++ {
 		bw.shutdownWG.Add(1)
-		go bw.execute(i)
+		go bw.runPoller()
 	}
+	bw.shutdownWG.Add(1)
+	go bw.runTaskDispatcher()
 
 	bw.isWorkerStarted = true
 	traceLog(func() {
-		bw.logger.Info("Started Worker", zap.Int("RoutineCount", bw.options.routineCount))
+		bw.logger.Info("Started Worker",
+			zap.Int("PollerCount", bw.options.pollerCount),
+			zap.Int("MaxConcurrentTask", bw.options.maxConcurrentTask),
+			zap.Int("MaxTaskRps", bw.options.maxTaskRps),
+		)
 	})
 }
 
+func (bw *baseWorker) isShutdown() bool {
+	select {
+	case <-bw.shutdownCh:
+		return true
+	default:
+		return false
+	}
+}
+
+func (bw *baseWorker) runPoller() {
+	defer bw.shutdownWG.Done()
+
+	bw.pollerRequestCh <- struct{}{} // initial poll request to kick off the poller loop
+
+	for {
+		select {
+		case <-bw.shutdownCh:
+			return
+		case <-bw.pollerRequestCh:
+			ch := make(chan struct{})
+			go func(ch chan struct{}) {
+				bw.pollTask()
+				close(ch)
+			}(ch)
+
+			// block until previous poll completed or return immediately when shutdown
+			select {
+			case <-bw.shutdownCh:
+				return
+			case <-ch:
+			}
+		}
+	}
+}
+
+func (bw *baseWorker) runTaskDispatcher() {
+	defer bw.shutdownWG.Done()
+
+	for i := 0; i < bw.options.maxConcurrentTask; i++ {
+		bw.taskSemaphore <- struct{}{}
+	}
+
+	for {
+		// wait for new task or shutdown
+		select {
+		case <-bw.shutdownCh:
+			return
+		case task := <-bw.taskQueueCh:
+			bw.pollerRequestCh <- struct{}{}
```

Comment:
I think starving pollers is fine, pollers need to do work only when workers are idle. And in the common case _pollLatency << processTaskLatency_, so I wouldn't worry about pre-fetching until we have data or a use-case that proves our assumption is wrong.  

To start with, we should shoot for correctness i.e never timeout tasks due to prefetching regardless of your worker configuration. We can always add prefetching support later, when we need it.

And even when we do prefetching, we probably will do it *smart* based on some smoothed out throughput estimation, just so that we can avoid unnecessary task timeouts.